pkgs/extension_discovery/README.md

```diff
@@ -0,0 +1,216 @@
+# Package Extension Discovery
+
+A convention to allow other packages to provide extension for your package.
+Including logic for finding extensions that honors this convention.
+
+The convention implemented in this package is that if `foo` provides an
+extension for `<targetPackage>`.
+Then `foo` must contain a config file `extension/<targetPackage>/config.json`.
+This file indicates that `foo` provides and extension for `<targetPackage>`.
+
+If `<targetPackage>` accepts extensions from other packages it must:
+ * Find extensions using `findExtensions('<targetPackage>')` from this package.
+ * Document how extensions are implemented:
+   * What should the contents of the `extension/<targetPackage>/config.json` file be?
+   * Should packages providing extension have a dependency constraint on `<targetPackage>`?
+   * What libraries/assets should packages that provide extension include?
+   * Should packages providing extensions specify a [topic in `pubspec.yaml`][1]
+     for easy discovery on pub.dev.
+
+The `findExtensions(targetPackage, packageConfig: ...)` function will:
+ * Load `.dart_tool/package_config.json` and find all packages that contain a
+   valid JSON file: `extension/<targetPackage>/config.json`.
+ * Provide the package name, location and contents of the `config.json` file,
+   for all detected extensions (aimed at `targetPackage`).
+ * Cache the results for fast loading, comparing modification timestamps to
+   ensure consistent results.
+
+It is the responsibility package that can be extended to validate extensions,
+decide when they should be enabled, and documenting how such extensions are
+created.
```

Comment:
should we add here that it is also the responsibility of the package that will be extended to detect when the cache has been invalidated so that they can call `findExtensions` again to get an updated set of extensions?

Reply:
If people want to watch the file-system for changes, then I think it makes sense to build it in here.

You'd need to watch `.dart_tool/package_config.json` and `config.json` for every mutable extension.

I imagine that devtools will load extensions when a debugging instance is launched. If I do `dart pub get` after launching my app, can I expect hotreloading of all dependencies to work?

I don't mind adding some watch logic.

Reply:
Adding some watch logic would be great. In the short term, we can probably get by with having a way for the user to manually refresh the list of extensions from the [extensions settings menu](https://github.com/flutter/devtools/pull/6138) and by refreshing the list of available extensions on each hot reload or restart event. This of course would only be possible for Flutter. If there was a running dart CLI app that DevTools was connected to, the app would need to be killed, restarted, and reconnected to DevTools to pick up any changes that happened from running `dart pub get`.

> I imagine that devtools will load extensions when a debugging instance is launched. If I do dart pub get after launching my app, can I expect hotreloading of all dependencies to work?

@christopherfujino do you know if all the dependencies are reloaded on hot reload? If we can expect this to work, DevTools can automatically refresh the list of extensions on hot reload / restart.